nalgebra-sparse/src/cs.rs

```diff
@@ -630,56 +631,63 @@ where
 
             // We test for in-bounds, uniqueness and monotonicity at the same time
             // to ensure that we only visit each minor index once
-            let mut iter = minor_idx_in_lane.iter();
             let mut prev = None;
             let mut nonmonotonic = false;
 
-            while let Some(next) = iter.next().copied() {
-                if next >= minor_dim {
+            for &minor_idx in minor_idx_in_lane {
+                if minor_idx >= minor_dim {
                     return Err(SparseFormatError::from_kind_and_msg(
                         SparseFormatErrorKind::IndexOutOfBounds,
                         "A minor index is out of bounds.",
                     ));
                 }
 
                 if let Some(prev) = prev {
-                    if prev == next {
+                    if prev == minor_idx {
                         return Err(SparseFormatError::from_kind_and_msg(
                             SparseFormatErrorKind::DuplicateEntry,
                             "Input data contains duplicate entries.",
                         ));
                     }
-                    if prev > next {
+                    if prev > minor_idx {
+                        if !sort {
+                            return Err(SparseFormatError::from_kind_and_msg(
+                                SparseFormatErrorKind::InvalidStructure,
+                                "Minor indices are not monotonically increasing within each lane.",
+                            ));
+                        }
                         nonmonotonic = true;
                     }
                 }
-                prev = Some(next);
+                prev = Some(minor_idx);
             }
 
             // sort if indices are nonmonotonic and sorting is expected
             if nonmonotonic && sort {
-                minor_index_permutation.resize(range_end - range_start, 0);
-                for (index, value) in (range_start..range_end).enumerate() {
-                    minor_index_permutation[index] = value;
+                let range_size = range_end - range_start;
+                minor_index_permutation.resize(range_size, 0);
+                compute_sort_permutation(&mut minor_index_permutation, &minor_idx_in_lane);
+                minor_idx_buffer.resize(range_size, 0);
+                for (index, &value) in minor_idx_in_lane.iter().enumerate() {
+                    minor_idx_buffer[index] = value;
```

Comment:
Couldn't this be simplified to
```rust
minor_idx_buffer.clear();
minor_idx_buffer.clone_from_slice(&minor_idx_in_lane);
```
?

Reply:
It does not work directly as `clear()` does not change the size of the buffer which might be required on each loop iteration here. But one can use `reserve()` additionally and push required values to the buffer. 
And it looks like with this approach we can get rid of using T::Zero indeed.

Reply:
Perhaps I misunderstand you, but if you use `clone_from_slice` as in my example you do not have to manage the resizing explicitly.

Reply:
I just tried to use exactly these two lines again and I get errors like `'unit_tests::csr::csr_matrix_valid_data_unsorted_column_indices' panicked at 'source slice length (3) does not match destination slice length (0)` when I run the unit tests.
Maybe I did something wrong but if we take a look into [documentation](https://doc.rust-lang.org/std/primitive.slice.html#method.clone_from_slice) , it looks like both slices have to be of same length.

Reply:
Sorry, I meant `extend_from_slice`!